plugins/com.google.cloud.tools.eclipse.preferences.test/src/com/google/cloud/tools/eclipse/preferences/areas/AreaPreferencePageTest.java

```diff
@@ -0,0 +1,240 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.eclipse.preferences.areas;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.jface.dialogs.IMessageProvider;
+import org.eclipse.jface.preference.IPersistentPreferenceStore;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Display;
+import org.eclipse.swt.widgets.Shell;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class AreaPreferencePageTest {
+  private @Mock IPersistentPreferenceStore preferences;
+
+  private Shell shell;
+
+  @After
+  public void tearDown() {
+    if (shell != null && !shell.isDisposed()) {
+      shell.dispose();
+    }
+  }
+
+  @Test
+  public void testWritesOnApply() {
+    // add an Area, call performApply, verify 1 call to preferences.store()
+    AreaBasedPreferencePage page = new AreaBasedPreferencePage("test");
+    page.addArea(new TestPrefArea("pref1", "value", preferences));
+    show(page);
+    assertTrue(page.performOk());
+    verify(preferences).setValue("pref1", "value");
+  }
+
+  @Test
+  public void testNoWritesOnCancel() {
+    // add an Area, call performCancel, verify no calls to preferences.store()
+    AreaBasedPreferencePage page = new AreaBasedPreferencePage("test");
+    page.addArea(new TestPrefArea("pref1", "value", preferences));
+    show(page);
+
+    page.performCancel();
+    verify(preferences, never()).setValue(anyString(), anyString());
+  }
+
+  @Test
+  public void testErrorStatusResults() {
+    // add an Area, trigger its VALUE property, cause its status to be error.
+    // verify showing error message
+    AreaBasedPreferencePage page = new AreaBasedPreferencePage("test");
+    TestPrefArea area = new TestPrefArea("pref1", "value", preferences);
+    page.addArea(area);
+
+    show(page);
+    assertTrue(page.isValid());
+    assertEquals(IMessageProvider.NONE, page.getMessageType());
+
+    area.status = new Status(IStatus.ERROR, "foo", "message");
+    area.fireValueChanged(TestPrefArea.IS_VALID, true, false);
+
+    assertFalse(page.isValid());
+    assertEquals(IMessageProvider.ERROR, page.getMessageType());
+    assertEquals("message", page.getMessage());
+
+    // change its status to be OK, trigger its VALUE property.
+    // verify no error message
+    area.status = Status.OK_STATUS;
```

Comment:
Could this be split into 2 tests? One for becoming invalid and one for becoming valid?